src/dbnode/storage/index.go

```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
 		state = asyncQueryExecState{
 			exhaustive: true,
 		}
-		deadline = start.Add(timeout)
-		wg       sync.WaitGroup
+		deadline      = start.Add(timeout)
+		wg            sync.WaitGroup
+		totalWaitTime time.Duration
 	)
 
 	// Create a cancellable lifetime and cancel it at end of this method so that
 	// no child async task modifies the result after this method returns.
 	cancellable := xresource.NewCancellableLifetime()
+	// always cancel to cleanup. this might be a no-op if it was already canceled by the background task to interrupt
+	// the workers.
 	defer cancellable.Cancel()
 
-	indexMatchingStartTime := time.Now()
-	var totalWaitTime time.Duration
+	stdCtx, ctxFound := ctx.GoContext()
+	if !ctxFound {
+		stdCtx = stdctx.Background()
+	}
+	select {
+	case <-stdCtx.Done():
+		return false, fmt.Errorf("caller canceled the query before it started")
```

Comment:
nit: Use a `var errQueryCancelledBeforeStarted = errors.New("....")` to the top of this file if this is a constant?

Reply:
just used the standard `index.ErrQueryCanceled`